json2junit/json2junit.go

```diff
@@ -0,0 +1,315 @@
+			data = append(c.b, data...)
+			// Reset the buffer.
+			c.b = c.b[:0]
+		}
+
+		// Unmarshal the JSON.
+		// If the JSON is invalid, just ignore it.
+		var jsonEntry jsonEntry
+		if err := json.Unmarshal(data, &jsonEntry); err == nil {
+			// Process the JSON entry.
+			if err := c.processJSONEntry(jsonEntry); err != nil {
+				return 0, fmt.Errorf("failed to process line: %v", err)
+			}
+		}
+
+		// Skip the newline.
+		b = b[i+1:]
+	}
+	return n, nil
+}
+
+// Close marks the end of the go test output.
+func (c *Converter) Close() error {
+	if err := c.closeXML(); err != nil {
+		return fmt.Errorf("failed to close XML: %v", err)
+	}
+	return nil
+}
+
+// jsonEntry is a single entry in the JSON file.
+type jsonEntry struct {
+	Time    time.Time
+	Action  string
+	Package string
+	Test    string
+	Output  string
+	Elapsed float64
+}
+
+// The following structs definitions are taken from
+// https://llg.cubic.org/docs/junit/.
+
+type junitTestSuite struct {
+	XMLName   xml.Name   `xml:"testsuite"`
+	Name      string     `xml:"name,attr"`
+	Tests     int        `xml:"tests,attr"`
+	Failures  int        `xml:"failures,attr"`
+	Skipped   int        `xml:"skipped,attr"`
+	Time      float64    `xml:"time,attr"`
+	Timestamp string     `xml:"timestamp,attr"`
+	SystemOut *systemOut `xml:",omitempty"`
+	TestCases []*junitTestCase
+}
+
+func (s junitTestSuite) testCase(name string) *junitTestCase {
+	for _, tc := range s.TestCases {
+		if tc.Name == name {
+			return tc
+		}
+	}
+	return nil
+}
+
+type systemOut struct {
+	XMLName xml.Name `xml:"system-out"`
+	Content []byte   `xml:",cdata"`
+}
+
+type junitTestCase struct {
+	XMLName   xml.Name     `xml:"testcase"`
+	Classname string       `xml:"classname,attr"`
+	Name      string       `xml:"name,attr"`
+	Time      float64      `xml:"time,attr"`
+	Result    *junitResult `xml:",omitempty"`
+	systemOut []byte       // Placeholder for the event output.
+}
+
+type junitResult struct {
+	XMLName xml.Name
+	Message string `xml:"message,attr"`
+	Content []byte `xml:",cdata"`
+}
+
+// suite returns the suite with the given name, or nil if there is none.
+func (c *Converter) suite(name string) (int, *junitTestSuite) {
+	for i, s := range c.suites {
+		if s.Name == name {
+			return i, s
+		}
+	}
+	return 0, nil
+}
+
+// processLine converts a single JSON test line.
+func (c *Converter) processJSONEntry(entry jsonEntry) error {
+	switch entry.Action {
+	case "start":
+		// start a new test suite.
+		if _, suite := c.suite(entry.Package); suite != nil {
+			return fmt.Errorf("duplicate start entry for %v", entry.Package)
+		}
+		suite := &junitTestSuite{
+			Name:      entry.Package,
+			Timestamp: entry.Time.Format(time.RFC3339),
+		}
+		c.suites = append(c.suites, suite)
+	case "run":
+		// start a new test case.
+		_, suite := c.suite(entry.Package)
+		if suite == nil {
+			return fmt.Errorf("no start entry for %v", entry.Package)
+		}
+		testCase := suite.testCase(entry.Test)
+		if testCase == nil {
```

Comment:
Why no error if there's a duplicate test start, like the error for duplicate suite start?

Reply:
Done